feat: Added async api docs (#174)
devrev/fern-api-docs: fern/docs/pages/interact_agent.mdx

---
title: "DevRev Agents Async API Cookbook"
description: "A step-by-step guide for using DevRev's asynchronous API for agent execution"
---

# DevRev agents async API cookbook

This cookbook provides a step-by-step guide to use the asynchronous API for DevRev agents. It assumes you've already created an agent using the DevRev agent builder platform.

## Table of contents

- [Introduction](#introduction).
- [Setting Up Your Webhook](#setting-up-your-webhook).
- [Making Async API Calls](#making-async-api-calls).
- [Handling Webhook Events](#handling-webhook-events).
- [Implementation Patterns](#implementation-patterns)
  - [Custom Applications](#custom-applications)
  - [Talk To Agent Node in Workflows](#talk-to-agent-node-in-workflows).
- [Troubleshooting](#troubleshooting).
- [Troubleshooting](#troubleshooting).

> ## Introduction
> 
> <Callout type="info">
>   The DevRev Async API enables long-running agent executions without timeout
>   concerns.
> </Callout>

Reviewer: Remove. It's not beneficial to start with an "Introduction" heading. The callout duplicates the first sentence of the paragraph.
```suggestion
```

This DevRev async API enables you to execute agents without any execution timeout concerns. The execution occurs as an asynchronous workflow that sends events to you via webhooks. This approach ensures:

- Agent complexity does not bring timeout concerns with it.
- Complete execution tracking via progress events.
- Ability to handle long-running agent tasks.

> **Disclaimer:** In any unlikely and unforeseen circumstances, if an individual operation of the workflow times out, you do not receive an error event right now to notify that you should not wait for any more messages. It is better to have a client side timeout for now while the issue is brainstormed and fixed for you. 

Reviewer: ```suggestion
<Callout type="note">
In any unlikely and unforeseen circumstances, if an individual operation of the workflow times out, you do not receive an error event right now to notify that you should not wait for any more messages. It is better to have a client side timeout for now while the issue is brainstormed and fixed for you. 
</Callout>
```

> ## Setting up your webhook

Reviewer: No gerund in title. https://www.notion.so/devrev/How-should-instructions-be-written-dac658b248d94db6a0c301079cffbbf5?pvs=4
```suggestion
## Set up webhook
```

Before using the async API, you need to create a webhook to receive agent execution events:

  ```bash
  curl --location 'https://api.devrev.ai/internal/webhooks.create' \
  --header 'Content-Type: application/json' \
  --header 'Authorization: <YOUR_API_TOKEN>' \
  --data '{
      "url": "https://your-application.com/webhook-endpoint",
      "event_types": ["ai_agent_response"],
      "headers": [
        {
          "name": "x-api-key",
          "value": "your-secret-key"
        }
      ]
  }'
  ```

>   ```javascript

Reviewer: You can use tabs for equivalent code in multiple languages. https://buildwithfern.com/learn/docs/content/components/code-blocks#code-blocks-with-tabs

  const axios = require('axios');

  async function createWebhook() {
    const response = await axios.post('https://api.devrev.ai/internal/webhooks.create', {
      url: "https://your-application.com/webhook-endpoint",
      event_types: ["ai_agent_response"],
      headers: [
        {
          name: "x-api-key",
          value: "your-secret-key"
        }
      ]
    }, {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': '<YOUR_API_TOKEN>'
      }
    });

    return response.data;
  }
  ```

Save the webhook ID from the response (format: `don:integration:dvrv-us-1:devo/0:webhook/auCJ7By8`). You need this when making async API calls.

Make sure that your webhook endpoint can appropriately respond to the verify events sent by DevRev to verify your webhook. This verify event has a challenge string which you need to return in response. You return the response like this:

  ```json
  200 OK
  {
    "challenge": "DlrVaK7zRyZWwbJhj5dZHDlrVaK7Jhj5dZZjH"
  }
  ```

You should follow the documentation provided for webhooks [here](https://developer.devrev.ai/public/guides/webhooks)

> ## Making async API calls

Reviewer: ```suggestion
## Make async API calls
```

    },
    "client_metadata": {
        "user_id": "12345",
        "conversation_id": "conv-789"
    }
  }'
  ```

  ```javascript
  const executeAgent = async () => {
    const response = await fetch('https://api.devrev.ai/internal/ai-agents.events.execute-async', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Authorization': '<YOUR_API_KEY>'
      },
      body: JSON.stringify({
        agent: "don:core:dvrv-us-1:devo/0:ai_agent/1",
        event: {
          input_message: {
            message: "Your query to the agent"
          }
        },
        session_object: "unique_conversation_identifier",
        webhook_target: {
          webhook: "don:integration:dvrv-us-1:devo/0:webhook/auCJ7By8"
        },
        client_metadata: {
          user_id: "12345",
          conversation_id: "conv-789"
        }
      })
    });
    
    return await response.json();
  };
  ```

<Callout type="warning">
  The session_object parameter is critical for maintaining conversation state
  across multiple calls.
</Callout>

## Handling webhook events

> Your webhook endpoint receives three types of events:

Reviewer: ```suggestion
Your webhook endpoint receives three types of events.
```

> ### 1. Progress messages

Reviewer: No number since it's not a sequence.
```suggestion
### Progress messages
```

These show which skills are being triggered and executed.

> #### Skill triggered
> 
>   ```json

Reviewer: ```suggestion
  ```json Skill triggered
```

  {
    "payload": {
      "ai_agent_response": {
        "agent": "don:core:dvrv-us-1:devo/xyz:ai_agent/123",
        "agent_response": "progress",
        "client_metadata": { /* your original metadata */ },
        "progress": {
          "progress_state": "skill_triggered",
          "skill_triggered": {
            "args": { /* skill arguments */ },
            "skill_name": "SkillName",
            "workflow": { /* This will not be populated if KnowledgeSearch skill is called */
              "display_id": "workflow-84",
              "id": "don:integration:dvrv-us-1:devo/xyz:workflow/84"
            }
          }
        },
        "session": "don:core:dvrv-us-1:devo/xyz:ai_agent_session/3810",
        "session_object": "unique_conversation_identifier"
      },
      "type": "ai_agent_response"
    }
  }
  ```

> #### Skill executed
> 
>   ```json

Reviewer: ```suggestion
  ```json Skill executed
```

      },
      "type": "ai_agent_response"
    }
  }
  ```

### 2. Final message

This event contains the agent's final response after all skills are executed.

```json
{
  "payload": {
    "ai_agent_response": {
      "agent": "don:core:dvrv-us-1:devo/xyz:ai_agent/123",
      "agent_response": "message",
      "client_metadata": {
        /* your original metadata */
      },
      "message": "The agent's final response message",
      "session": "don:core:dvrv-us-1:devo/xyz:ai_agent_session/3810",
      "session_object": "unique_conversation_identifier"
    },
    "type": "ai_agent_response"
  }
}
```

### 3. Error Message

This event indicates an error occurred during execution.

```json
{
  "payload": {
    "ai_agent_response": {
      "agent": "don:core:dvrv-us-1:devo/xyz:ai_agent/123",
      "agent_response": "error",
      "client_metadata": {
        /* your original metadata */
      },
      "error": {
        "error": "Error description"
      },
      "session": "don:core:dvrv-us-1:devo/xyz:ai_agent_session/3810",
      "session_object": "unique_conversation_identifier"
    },
    "type": "ai_agent_response"
  }
}
```

## Implementation patterns

> ### Custom Applications

Reviewer: ```suggestion
### Custom applications
```

> For building custom applications with the async API:

Reviewer: Move after callout.

      event: {
        input_message: {
          message: message,
        },
      },
      session_object: "conversation_" + userId,
      webhook_target: {
        webhook: "your_webhook_id",
      },
      client_metadata: {
        conversation_id: conversationId,
      },
    }),
  });
}

// 2. Webhook handler (server-side)
app.post("/webhook-endpoint", (req, res) => {
  const event = req.body;

  // Add handling for verify events

  const conversationId =
    event.payload.ai_agent_response.client_metadata.conversation_id;

  if (event.payload.ai_agent_response.agent_response === "message") {
    // Final message
    const message = event.payload.ai_agent_response.message;
    sendToClient(conversationId, {
      type: "agent_response",
      message: message,
    });
  } else if (event.payload.ai_agent_response.agent_response === "error") {
    // Error occurred
    sendToClient(conversationId, {
      type: "agent_error",
      error: event.payload.ai_agent_response.error.error,
    });
  }

  res.status(200).send("OK");
});
```

<Callout type="tip">
  Using WebSockets or Server-Sent Events can provide real-time updates to your
  UI as events are received.
</Callout>

> ### Talk To Agent Node in Workflows

Reviewer: ```suggestion
### Talk To agent node in workflows
```

> **Note:** This is in early access, please contact support to have it enabled for you.

Reviewer: ```suggestion
<Callout type="note">
This functionality is in early access. Contact support to have it enabled for you.
</Callout>
```
